include/forcing/ForcingsEngineDataProvider.hpp

```diff
@@ -0,0 +1,258 @@
+
+  protected:
+    using storage_type = detail::ForcingsEngineStorage;
+
+    ForcingsEngineDataProvider(
+        std::string init,
+        std::size_t time_begin_seconds,
+        std::size_t time_end_seconds
+    )
+      : time_begin_(std::chrono::seconds{time_begin_seconds})
+      , time_end_(std::chrono::seconds{time_end_seconds})
+      , bmi_(storage_type::instances.get(init))
+      , time_step_(std::chrono::seconds{static_cast<int64_t>(bmi_->GetTimeStep())})
+      , time_current_index_(std::chrono::seconds{static_cast<int64_t>(bmi_->GetCurrentTime())} / time_step_)
+      , var_output_names_(bmi_->GetOutputVarNames())
+    {}
+
+    
+    //! Update the Forcings Engine instance to the next timestep.
+    void next()
+    {
+        bmi_->Update();
+        time_current_index_++;
+    }
+
+    //! Update the Forcings Engine instance to the next timestep before `time`.
+    //! @param time Time in seconds to update to.
+    //!             i.e. A value of 14401 will update
+    //!                  the instance to the 5th time index,
+    //!                  since 3600 * 4 = 14400 but 14400 < 14401.
+    void next(double time)
+    {
+      const auto start = bmi_->GetCurrentTime();
+      bmi_->UpdateUntil(time);
+      const auto end = bmi_->GetCurrentTime();
+      time_current_index_ += static_cast<int64_t>(
+        (end - start) / bmi_->GetTimeStep()
+      );
+    }
+
+    //! Forcings Engine instance
+    std::shared_ptr<models::bmi::Bmi_Py_Adapter> bmi_ = nullptr;
+
+    //! Output variable names
+    std::vector<std::string> var_output_names_{};
+
+  private:
+    //! Initialization config file path
+    std::string init_;
+
+    clock_type::time_point time_begin_{};
+    clock_type::time_point time_end_{};
+    clock_type::duration   time_step_{};
+    std::size_t            time_current_index_{};
+};
+
+
+//! Forcings Engine factory constructor function
+//! @param init File path to instance initialization file.
+//! @param time_begin Beginning time of simulation in timestamp form.
+//! @param time_end End time of simulation in timestamp form.
+//! @param time_fmt Parse format for timestamp strings.
+//! @param args Additional arguments passed to derived constructor.
+//! @tparam Tp Derived Forcings Engine Type.
+//! @tparam Args Additional argument types passed to derived constructor.
+//! @note This function requires that @c Tp must derive from
+//!       type @c ForcingsEngineDataProvider<Tp::data_type, Tp::selection_type>.
+template<
+  typename Tp,
+  typename... Args,
+  std::enable_if_t<
+    std::is_base_of<
+      ForcingsEngineDataProvider<
+        typename Tp::data_type,
+        typename Tp::selection_type
+      >,
+      Tp
+    >::value,
+    bool
+  > = true
+>
+std::unique_ptr<typename Tp::base_type> make_forcings_engine(
+  const std::string& init,
+  const std::string& time_begin,
+  const std::string& time_end,
+  const std::string& time_fmt = default_time_format,
+  Args&&...          args
+)
```

Comment:
It probably makes more sense to just return `std::unique_ptr<Tp>`, and let polymorphism -- if needed for some reason -- be handled on assignment by consumers of this function.

Reply:
Now that I'm taking the 100th look at this, but with a clearer head, I'm pretty sure this isn't needed.

- The instance getting/construction logic can be moved to the protected constructor
- The assertion is planned to move outside of this to a single call point
- All that's left is `std::make_unique<Tp>`, which doesn't require a convenience function, and on its own removes the need for any form of type checking.